### PR TITLE
clangcuda fixes

### PR DESCRIPTION
**Pull request by MrBurmark**
Fixes to enable compilation with clang cuda compiler.
The clang cuda compiler at this point does not generate correct code for __shfl intrinsics so all cuda reducer tests fail, and some RAJA kernel tests using cuda fail with an invalid device function error.

Features:
  example script and host-config for clangcuda configuration
  non-atomic host implementations for atomics using the cuda_atomic policy.
  update tpl/cub to 1.8.0
  guard __shfl_sync intrinsics with clangcuda compatible macro
  add __shfl_sync wrapper specializations for supported types

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
This is awesome!

**Comment by DavidPoliakoff** [comment]
Oh this is great. Thanks!

**Comment by davidbeckingsale** [comment]
Did you want to fix the reductions stuff prior to merging, or is that impossible atm?

**Comment by MrBurmark** [comment]
I don't think its worth working around clangcuda for the reducers.

**Comment by rhornung67** [comment]
@MrBurmark do the reducers wirk with cuda 9.x?

**Comment by MrBurmark** [comment]
No the code generated by all of the clangcuda versions that I've tried is incorrect.

**Comment by rhornung67** [comment]
Grrrrfffff....

**Comment by DavidPoliakoff** [comment]
I can't really document what's going on. Would anybody be game to file a bug here https://bugs.llvm.org/enter_bug.cgi , or get in touch with Chris Earl?

**Comment by DavidPoliakoff** [comment]
Excuse the multiple messages, but I sent an email to Chris, I recommend we discuss there

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,6 +48,7 @@ if (NOT RAJA_LOADED)
   option(ENABLE_CLANG_CUDA "Use Clang's native CUDA support" Off)
   set(CUDA_ARCH "sm_35" CACHE STRING "Compute architecture to pass to CUDA builds")
   option(ENABLE_TESTS "Build tests" On)
+  option(ENABLE_ISSUES "Build issue reproducers" Off)
```
Comment (davidbeckingsale) [inline CMakeLists.txt]: I think this is a good idea, but I propose we rename it "reproducers".

File: issues/clangcuda/issue-shuffle.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
+///
+/// Source file containing clangcuda __shfl/__shfl_sync issue reproducer.
+///
+/// Immediate Issue: Warp reductions using shuffle intrinsics give wrong results.
+///
+/// Larger Issue: RAJA reducer tests fail.
+///
+/// Seen when compiling for sm_60 on Sierra EA systems (IBM p8, Nvidia P100)
+/// Seen when compiling for sm_70 on Sierra systems (IBM p9, Nvidia V100)
```
Comment (davidbeckingsale) [inline issues/clangcuda/issue-shuffle.cpp]: Does this reproduce on TOSS3?

Comment (MrBurmark) [reply issues/clangcuda/issue-shuffle.cpp]: It didn't reproduce on hasgpu when I tried it.

Comment (davidbeckingsale) [reply issues/clangcuda/issue-shuffle.cpp]: That's interesting. Thanks for the update. Have you tried on pascal?

Comment (MrBurmark) [reply issues/clangcuda/issue-shuffle.cpp]: Yes, it does reproduce on manta. I didn't put up a host config as I'm using my own build of clang trunk

**Comment by davidbeckingsale** [comment]
I meant the machine that has pascals with Intel CPUs :)

**Comment by MrBurmark** [comment]
I haven't tried it on that machine.

**Comment by davidbeckingsale** [comment]
Do we want to go ahead and merge this?

**Comment by rhornung67** [comment]
@davidbeckingsale it looks ok to me, but the compiler still has issues. @MrBurmark what do you think? Should we merge this?

**Comment by MrBurmark** [comment]
I'm fine with merging this. Could I get an approve?

**Comment by davidbeckingsale** [comment]
I think it's good to get these changes in so they don't go stale, and then we can continue to work on getting the reducers fixed.

File: host-configs/lc-builds/toss3/clangcuda_6_0_0_nvcc_8_0.cmake [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+##
+## For details about use and distribution, please read RAJA/LICENSE.
+##
+
+set(RAJA_COMPILER "RAJA_COMPILER_CLANG" CACHE STRING "")
+
+set(CMAKE_CXX_COMPILER "/usr/tce/packages/clang/clang-6.0.0/bin/clang++" CACHE PATH "")
+set(CMAKE_C_COMPILER "/usr/tce/packages/clang/clang-6.0.0/bin/clang" CACHE PATH "")
+
+set(CMAKE_EXE_LINKER_FLAGS "-L/usr/tce/packages/cuda/cuda-8.0/lib64 -lcudart_static -lcudadevrt -lrt -ldl -lnvToolsExt -pthread -Wl,-rpath=/usr/tce/packages/clang/clang-6.0.0/lib" CACHE PATH "")
```
[the remark this answers is not in the document]
Comment (MrBurmark) [reply host-configs/lc-builds/toss3/clangcuda_6_0_0_nvcc_8_0.cmake]: Clang 6 is only compatible with up to cuda 8.
Clang tip of trunk is cuda 9 compatible and did reproduce the issue.

**Pull request merged**